### PR TITLE
Elasticsearch 6 changes

### PR DESCRIPTION
**Pull request by nitishgoyal13**

### CODE REVIEW

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/AnalyticsLoader.java [hunk 1]
```diff
@@ -60,17 +61,19 @@
     private final CacheManager cacheManager;
     private final ObjectMapper objectMapper;
     private final EmailConfig emailConfig;
+    private EmailClient emailClient;
 
     public AnalyticsLoader(TableMetadataManager tableMetadataManager, DataStore dataStore, QueryStore queryStore,
                            ElasticsearchConnection elasticsearchConnection, CacheManager cacheManager, ObjectMapper objectMapper,
-                           EmailConfig emailConfig) {
+                           EmailConfig emailConfig, EmailClient emailClient) {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/AnalyticsLoader.java]: Why is analyticsloader and actionexecutor interface polluted with email stuff? Remove both and create a listener interface for exception. Implement that to send email. Add loader to register listeners to action.

File: foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/impl/DistributedCacheTest.java [hunk 2]
```diff
@@ -64,8 +65,15 @@ public void setUp() throws Exception {
         when(tableMetadataManager.exists(TestUtils.TEST_TABLE_NAME)).thenReturn(true);
         QueryStore queryStore = Mockito.mock(QueryStore.class);
 
+        EmailConfig emailConfig = new EmailConfig();
```
Comment (santanusinha) [inline foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/impl/DistributedCacheTest.java]: This is not needed. Mock things properly.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java [hunk 3]
```diff
@@ -83,7 +83,11 @@
 
     public GroupAction(GroupRequest parameter, String cacheToken, AnalyticsLoader analyticsLoader) {
         super(parameter, cacheToken, analyticsLoader);
-        emailClient = getEmailClient(analyticsLoader.getEmailConfig());
+        if (analyticsLoader.getEmailClient() != null){
+            emailClient = analyticsLoader.getEmailClient();
+        }else {
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java]: Formatting issue. This is the worst possible place to do this. Email can be sent for a variety of reasons. Please have either use:
- a Guava MessageBus to send out events on which send out mails from a subscriber
- create a listener interface and implement that to send email.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/ActionMetadata.java [hunk 4]
```diff
@@ -28,6 +29,7 @@
     private final Class<? extends Action> action;
     private final boolean cacheable;
     private final String cacheToken;
+    private EmailClient emailClient;
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/ActionMetadata.java]: This is metadata. Why is EmailClient being added here?

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/AnalyticsLoader.java [hunk 5]
```diff
@@ -60,17 +61,19 @@
     private final CacheManager cacheManager;
     private final ObjectMapper objectMapper;
     private final EmailConfig emailConfig;
+    private EmailClient emailClient;
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/AnalyticsLoader.java]: Remove this.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/DistributedTableMetadataManager.java [hunk 6]
```diff
@@ -639,7 +639,7 @@ private void saveCardinalityCache(String table, TableFieldMapping tableFieldMapp
                     .execute()
                     .get(2, TimeUnit.SECONDS);
         } catch (Exception e) {
-            logger.error("Error in saving cardinality cache: " + e.getMessage(), e);
+            logger.error("Error in saving cardinality cache: ");
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/table/impl/DistributedTableMetadataManager.java]: Print the error.

File: foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/QueryExecutorTest.java [hunk 7]
```diff
@@ -76,9 +77,15 @@ public void setUp() throws Exception {
         when(tableMetadataManager.exists(anyString())).thenReturn(true);
         when(tableMetadataManager.get(anyString())).thenReturn(TestUtils.TEST_TABLE);
         QueryStore queryStore = mock(QueryStore.class);
+        EmailConfig emailConfig = new EmailConfig();
```
Comment (santanusinha) [inline foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/QueryExecutorTest.java]: Remove this and mock properly

File: foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/ActionTest.java [hunk 8]
```diff
@@ -64,30 +68,18 @@ public void setUp() throws Exception {
         Config config = new Config();
         EmailConfig emailConfig = new EmailConfig();
         emailConfig.setHost("127.0.0.1");
+        emailConfig.setFrom("noreply@foxtrot.com");
+        EmailClient emailClient = Mockito.mock(EmailClient.class);
```
Comment (santanusinha) [inline foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/ActionTest.java]: Why is config needed if EmailClient is being mocked?

File: foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java [hunk 9]
```diff
@@ -779,7 +779,7 @@ public void testMissingIndicesQuery() throws FoxtrotException {
                 .indices()
                 .getIndex(new GetIndexRequest())
                 .actionGet();
-        assertEquals(3, Arrays.stream(response.getIndices())
+        assertEquals(4, Arrays.stream(response.getIndices())
```
Comment (santanusinha) [inline foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java]: Why is this change needed now? The root cause needs to be checked. Test assertions should not be changed to make tests pass. The whole reason of assertions is to catch issues. If the assertion itself is changed, then there is no point of writing tests by the original authors.

File: foxtrot-server/src/main/java/com/flipkart/foxtrot/server/FoxtrotServer.java [hunk 10]
```diff
@@ -184,7 +151,7 @@ public void run(FoxtrotServerConfiguration configuration, Environment environmen
         FoxtrotTableManager tableManager = new FoxtrotTableManager(tableMetadataManager, queryStore, dataStore);
         CacheManager cacheManager = new CacheManager(new DistributedCacheFactory(hazelcastConnection, objectMapper, cacheConfig));
         AnalyticsLoader analyticsLoader = new AnalyticsLoader(tableMetadataManager, dataStore, queryStore, elasticsearchConnection,
-                                                              cacheManager, objectMapper, emailConfig
+                                                              cacheManager, objectMapper, emailConfig, null
```
Comment (santanusinha) [inline foxtrot-server/src/main/java/com/flipkart/foxtrot/server/FoxtrotServer.java]: This will change as mentioned earlier.

File: foxtrot-server/src/test/java/com/flipkart/foxtrot/server/resources/FoxtrotResourceTest.java [hunk 11]
```diff
@@ -152,8 +136,15 @@ public FoxtrotResourceTest() throws Exception {
         queryStore = new ElasticsearchQueryStore(tableMetadataManager, elasticsearchConnection, dataStore, mapper, cardinalityConfig);
         queryStore = spy(queryStore);
 
+        EmailConfig emailConfig = new EmailConfig();
+        emailConfig.setHost("127.0.0.1");
+        emailConfig.setFrom("noreply@foxtrot.com");
+        EmailClient emailClient = Mockito.mock(EmailClient.class);
+        when(emailClient.sendEmail(any(String.class), any(String.class), any(String.class))).thenReturn(true);
```
Comment (santanusinha) [inline foxtrot-server/src/test/java/com/flipkart/foxtrot/server/resources/FoxtrotResourceTest.java]: Same issues as before

File: pom.xml [hunk 12]
```diff
@@ -255,26 +229,6 @@
                     <target>1.8</target>
                 </configuration>
             </plugin>
-            <plugin>
```
Comment (santanusinha) [inline pom.xml]: Do not remove jacoco. It is used to calculate code coverage.

File: foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java (hunk 9 above)
Comment (nitishgoyal13) [reply foxtrot-core/src/test/java/com/flipkart/foxtrot/core/querystore/actions/FilterActionTest.java]: We have a new index for cardinality being created. So, number of default indices increased from 3 to 4

File: pom.xml (hunk 12 above)
Comment (nitishgoyal13) [reply pom.xml]: This was there twice in the pom. I removed one with older version

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/AnalyticsLoader.java (hunk 1 above)
Comment (nitishgoyal13) [reply foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/AnalyticsLoader.java]: Will make these changes

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java (hunk 3 above)
Comment (nitishgoyal13) [reply foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/GroupAction.java]: Will make these changes

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/ActionMetadata.java (hunk 4 above)
Comment (nitishgoyal13) [reply foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/ActionMetadata.java]: This is removed

**Comment by santanusinha** [comment]
The test is to check for test indices.  Rewriting the match to achieve the
same should have been sufficient. Otherwise this line will keep changing
for every meta index addition.
